src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,719 @@
+
+  *buff_ptr++ = keep_alive_msb;
+  *buff_ptr++ = keep_alive_lsb;
+
+  buff_ptr = iotjs_mqtt_string_serialize(buff_ptr, client_id);
+
+  jerry_release_value(jclient_id);
+  jerry_release_value(jusername);
+  jerry_release_value(jpassword);
+
+  return jbuff;
+}
+
+
+JS_FUNCTION(MqttPublish) {
+  DJS_CHECK_THIS();
+
+  DJS_CHECK_ARGS(1, object);
+
+  jerry_value_t joptions = JS_GET_ARG(0, object);
+
+  jerry_value_t jmessage =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_MESSAGE);
+  jerry_value_t jtopic =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_TOPIC);
+  jerry_value_t jretain =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_RETAIN);
+  jerry_value_t jqos =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_QOS);
+  jerry_value_t jpacket_id =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_PACKETID);
+
+  uint8_t qos = 0;
+  if (!jerry_value_is_undefined(jqos)) {
+    qos = jerry_get_number_value(jqos);
+  }
+
+  bool dup = false;
+
+  uint8_t header_byte = 0;
+  header_byte |= (PUBLISH << 4);
+  header_byte |= (dup << 3);
+  header_byte |= (qos << 1);
+  header_byte |= (jerry_get_boolean_value(jretain));
+
+  iotjs_bufferwrap_t *message_payload = iotjs_bufferwrap_from_jbuffer(jmessage);
+  iotjs_bufferwrap_t *topic_payload = iotjs_bufferwrap_from_jbuffer(jtopic);
+
+  uint8_t packet_identifier_lsb = 0;
+  uint8_t packet_identifier_msb = 0;
+
+  // IF NOT QOS
+  if (!jerry_value_is_undefined(jqos) &&
+      !jerry_value_is_undefined(jpacket_id) &&
+      jerry_value_is_number(jpacket_id)) {
+    uint16_t packet_identifier = jerry_get_number_value(jpacket_id);
+    packet_identifier_msb = (uint8_t)(packet_identifier >> 8);
+    packet_identifier_lsb = (uint8_t)(packet_identifier & 0x00FF);
+  }
+
+  size_t payload_len = message_payload->length + IOTJS_MQTT_LSB_MSB_SIZE;
+  size_t variable_header_len = topic_payload->length + IOTJS_MQTT_LSB_MSB_SIZE +
+                               sizeof(packet_identifier_msb) +
+                               sizeof(packet_identifier_lsb);
+  uint32_t remaining_length = payload_len + variable_header_len;
+  size_t full_len = sizeof(header_byte) +
+                    get_remaining_length_size(remaining_length) +
+                    variable_header_len + payload_len;
+
+  jerry_value_t jbuff = iotjs_bufferwrap_create_buffer(full_len);
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jbuff);
+
+  uint8_t *buff_ptr = (uint8_t *)buffer_wrap->buffer;
+
+  *buff_ptr++ = header_byte;
+  buff_ptr = iotjs_encode_remaining_length(remaining_length, buff_ptr);
+  buff_ptr = iotjs_mqtt_string_serialize(buff_ptr, topic_payload);
+  *buff_ptr++ = packet_identifier_msb;
+  *buff_ptr++ = packet_identifier_lsb;
+
+  buff_ptr = iotjs_mqtt_string_serialize(buff_ptr, message_payload);
+
+  jerry_release_value(jmessage);
+  jerry_release_value(jtopic);
+  jerry_release_value(jretain);
+
+  return jbuff;
+}
+
+
+JS_FUNCTION(MqttHandle) {
```

Comment:
What happens if the package is not fully arrived?

Reply:
The maximum TCP package size is 64K, but due to memory limitations the IoT.js can't handle such a big packages.